File: 332_reconstruct_ltinerary.py
import collections

class Solution(object):
  def findItinerary(self,tickets):
    mymap = collections.defaultdict(list)
    res = []
    n = len(tickets)
    for ticket in tickets:
      mymap[ticket[0]] += [ticket[1]]
    get = self.dfs('JFK',res,mymap,0,n)
    return get == True and res

  def dfs(self,cur,res,mymap,i,n):
    if len(res) == n: 
      res.append(cur)
      return True;
    d = sorted(mymap[cur])
    for des in d:
        res.append(cur)
        # important: should remove what you add!
        mymap[cur].remove(des)
        get = self.dfs(des,res,mymap,i+1,n)
        if not get:
          mymap[cur].append(des)
          res.pop(-1)
        else:
          return True
    

      

solution = Solution()
tickets = [["JFK","SFO"],["JFK","ATL"],["SFO","ATL"],["ATL","JFK"],["ATL","SFO"]]

print solution.findItinerary(tickets)

File: 332_reconstruct_ltinerary.cpp
#include <cstdio>
#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <set>
#include <unordered_map>
#include <unordered_set>
#include <queue>
#include <algorithm>

#define INT_MAX 0x7fffffff
#define INT_MIN 0x80000000

using namespace std;

void dfs(string cur,vector<string> &res, unordered_map<string,multiset<string>> &my_map){
  while(my_map[cur].size()){
    string next = *got->second.begin();
    got->second.erase(got->second.begin());
    dfs(next,res,my_map);  
  }
  res.push_back(cur);
}

vector<string> findItinerary(vector<pair<string,string>> tickets){
  unordered_map<string, multiset<string>> my_map;
  vector<string> res;
  string init = "JFK";
  for(auto ticket : tickets){
    my_map[ticket.first].insert(ticket.second); 
  }
  dfs(init,res,my_map);
  reverse(res.begin(),res.end());
  return res;
}


int main(){

}
